core/src/main/scala/kafka/log/FileMessageSet.scala

```diff
@@ -149,6 +155,67 @@ class FileMessageSet private[kafka](@volatile var file: File,
   }
 
   /**
+   * Search forward for the message whose timestamp is greater than or equals to the target timestamp.
+   *
+   * The search will stop immediately when it sees a message in format version before 0.10.0. This is to avoid
+   * scanning the entire log when all the messages are still in old format.
+   *
+   * @param targetTimestamp The timestamp to search for.
+   * @param startingPosition The starting position to search.
+   * @return None, if no message exists at or after the starting position.
+   *         Some(TimestampOffset) with (largest_timestamp_checked, the_next_offset_to_read) otherwise.
+   */
+  def searchForTimestamp(targetTimestamp: Long, startingPosition: Int): Option[TimestampOffset] = {
+    var maxTimestampChecked = Message.NoTimestamp
+    var lastOffsetChecked = -1L
+    val messagesToSearch = read(startingPosition, sizeInBytes)
+    for (messageAndOffset <- messagesToSearch) {
+      val message = messageAndOffset.message
+      lastOffsetChecked = messageAndOffset.offset
+      // Stop searching once we see message format before 0.10.0.
+      // This equivalent as treating message without timestamp has the largest timestamp.
+      // We do this to avoid scanning the entire log if no message has a timestamp.
+      if (message.magic == Message.MagicValue_V0)
+        return Some(TimestampOffset(maxTimestampChecked, messageAndOffset.offset))
```

Comment:
1. slotFor also returns -1 if the timestamp is smaller than the first indexed timestamp. So the case I described is something like the following: Say we have 1 GB log segment, the first message w/ a timestamp is offset 1234567 at position 900MB. And the timestamp is 1000. So the first time index entry will be (1000, 1234567) in this case. Now if user come and search for timestamp 100, ```indexSlotFor``` will return -1, and currently we will return baseOffset if ```indexSlotFor``` returns -1. So we will scan the log from base offset and page 900 MB into memory.

If we return the offset in the first time index entry. We don't need to scan the previous 900 MB in that segment. Because we know that there is no timestamp before the offset of the first indexed offset in the time index, there is no need to scan that.

In another scenario, if the time index is empty, that means there is no timestamp in the segment. But in that case ```LogSegment.largestTimestamp()``` will return the last modification time. So ```Log.fetchOffsetsByTimestamp()``` will not skip the segment but treat it the same way as the segments that have timestamps. This essentially means at Log level, we people search for we do not distinguish between the largestTimestamp from the message timestamp or the largestTimestamp from last modification time. They provides a unified experience to the users.

Are you suggesting that we always ignore the old messages completely in the new way of searching for timestamp. I am worried about the transition period regarding that approach. For example, say some user is now using the timestamp search at segment granularity. If they want to consume all the data after 8:00 AM, they can start to consume the data after the segment whose last modification time is before 8:00 AM. There may be duplicates but it works. Now the users upgraded to new consumer which search for timestamp based on message timestamp, suddenly we skip all the old segments and they always get log end offsets.

That was also why I am not sure if we should skip the old messages or we should just stop on seeing old messages.

2. I agree the case I described would be rare, so scanning until we find timestamp should be fine.